src/ssl.c

```diff
@@ -31951,8 +32090,101 @@ int wolfSSL_DSA_do_verify_ex(const unsigned char* digest, int digest_len,
     return WOLFSSL_SUCCESS;
 }
 #endif /* !HAVE_SELFTEST && !HAVE_FIPS */
-#endif /* NO_DSA */
 
+WOLFSSL_API int wolfSSL_i2d_DSAparams(const WOLFSSL_DSA* dsa,
+    unsigned char** out)
+{
+    int ret = 0;
+    word32 derLen = 0;
+    int increment = 1;
+    DsaKey* key = NULL;
+
+    WOLFSSL_ENTER("wolfSSL_i2d_DSAparams");
+
+    if (dsa == NULL || dsa->internal == NULL || out == NULL) {
+        ret = BAD_FUNC_ARG;
+    }
+
+    if (ret == 0) {
+        key = (DsaKey*)dsa->internal;
+        ret = wc_DsaKeyToParamsDer_ex(key, NULL, &derLen);
+        if (ret == LENGTH_ONLY_E) {
+            ret = 0;
+        }
+    }
+    if (ret == 0 && *out == NULL) {
+        /* If we're allocating out for the caller, we don't increment out just
+           past the end of the DER buffer. If out is already allocated, we do.
+           (OpenSSL convention) */
+        increment = 0;
+        *out = XMALLOC(derLen, key->heap, DYNAMIC_TYPE_OPENSSL);
+        if (*out == NULL) {
+            ret = MEMORY_E;
+        }
+    }
+    if (ret == 0) {
+        ret = wc_DsaKeyToParamsDer_ex(key, *out, &derLen);
+    }
+    if (ret >= 0 && increment == 1) {
+        *out += derLen;
+    }
```

Comment:
On error case should the allocated pointer be free'd?

Reply:
Yes, it should be. Good catch.